File: ai_houkai/eval.py
"""Retrieval-quality metrics and a tiny evaluation harness.

Dependency-free (stdlib only), matching the project's local-first philosophy.
Lets you measure ranking changes against a small gold set so retrieval tweaks
(weights, fusion, diversity, …) are no longer flying blind.

    from ai_houkai.eval import EvalCase, evaluate

    cases = [
        EvalCase(query="how do we deploy", relevant_ids=[deploy_mem.id]),
        EvalCase(query="test isolation",   relevant_ids=[a.id, b.id], k=3),
    ]
    result = evaluate(store, cases, default_mode="hybrid")
    print(result.recall_at_k, result.mrr, result.ndcg_at_k)

The metric functions also work standalone on any ranked list of ids:

    recall_at_k(["a", "b", "c"], relevant={"b"}, k=2)   # -> 1.0
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Iterable, Sequence

__all__ = [
    "recall_at_k", "precision_at_k", "reciprocal_rank", "average_precision",
    "dcg_at_k", "ndcg_at_k", "EvalCase", "EvalResult", "evaluate",
]


def recall_at_k(retrieved: Sequence[str], relevant: Iterable[str], k: int) -> float:
    """Fraction of relevant ids that appear in the top-k retrieved.

    Counts *distinct* matches so a duplicated retrieved id can't push the value
    above 1.0.
    """
    rel = set(relevant)
    if not rel:
        return 0.0
    top = set(retrieved[:k])
    return len(top & rel) / len(rel)


def precision_at_k(retrieved: Sequence[str], relevant: Iterable[str], k: int) -> float:
    """Fraction of the top-k retrieved that are relevant."""
    rel = set(relevant)
    top = list(retrieved[:k])
    if not top:
        return 0.0
    return sum(1 for r in top if r in rel) / len(top)


def reciprocal_rank(retrieved: Sequence[str], relevant: Iterable[str]) -> float:
    """1 / rank of the first relevant id (0 if none retrieved)."""
    rel = set(relevant)
    for i, r in enumerate(retrieved, start=1):
        if r in rel:
            return 1.0 / i
    return 0.0


def average_precision(retrieved: Sequence[str], relevant: Iterable[str]) -> float:
    """Average precision over the retrieved list (basis of MAP).

    Each relevant id is credited once, at its first occurrence.
    """
    rel = set(relevant)
    if not rel:
        return 0.0
    seen: set[str] = set()
    hits = 0
    total = 0.0
    for i, r in enumerate(retrieved, start=1):
        if r in rel and r not in seen:
            seen.add(r)
            hits += 1
            total += hits / i
    return total / len(rel)


def dcg_at_k(retrieved: Sequence[str], relevant: Iterable[str], k: int) -> float:
    """Discounted cumulative gain (binary relevance) over the top-k.

    A relevant id is credited once (at its first occurrence) so duplicate
    retrieved ids cannot inflate the gain.
    """
    rel = set(relevant)
    seen: set[str] = set()
    total = 0.0
    for i, r in enumerate(retrieved[:k], start=1):
        if r in rel and r not in seen:
            seen.add(r)
            total += 1.0 / math.log2(i + 1)
    return total


def ndcg_at_k(retrieved: Sequence[str], relevant: Iterable[str], k: int) -> float:
    """Normalised DCG@k (binary relevance), in [0, 1]."""
    rel = set(relevant)
    if not rel:
        return 0.0
    ideal_hits = min(len(rel), k)
    idcg = sum(1.0 / math.log2(i + 1) for i in range(1, ideal_hits + 1))
    if idcg == 0:
        return 0.0
    return dcg_at_k(retrieved, relevant, k) / idcg


@dataclass
class EvalCase:
    """One query with its known-relevant memory ids.

    ``k``/``mode`` are ``None`` by default so they fall back to the
    ``default_k``/``default_mode`` passed to :func:`evaluate`; set them per-case
    only to override.
    """
    query: str
    relevant_ids: list[str]
    k: int | None = None
    mode: str | None = None


@dataclass
class EvalResult:
    n: int
    k: int
    recall_at_k: float
    precision_at_k: float
    mrr: float
    map: float
    ndcg_at_k: float
    per_case: list[dict] = field(default_factory=list)

    def summary(self) -> str:
        ks = "mixed" if self.k < 0 else str(self.k)
        return (f"n={self.n} | recall@{ks}={self.recall_at_k:.3f} "
                f"P@{ks}={self.precision_at_k:.3f} MRR={self.mrr:.3f} "
                f"MAP={self.map:.3f} nDCG@{ks}={self.ndcg_at_k:.3f}")


def evaluate(
    store,
    cases: Sequence[EvalCase],
    *,
    default_k: int = 5,
    default_mode: str = "hybrid",
    **recall_kwargs,
) -> EvalResult:
    """Run each case through ``store.recall`` and aggregate ranking metrics.

    Recall is invoked read-only (``touch=False``) so evaluating does not perturb
    access-count / recency. Extra keyword args (e.g. ``weights=``, ``fusion=``,
    ``diversity=``) are forwarded to ``recall`` so you can A/B ranking configs.
    """
    rs = ps = rr = ap = ng = 0.0
    per: list[dict] = []
    ks_used: set[int] = set()
    for c in cases:
        k = c.k if getattr(c, "k", None) is not None else default_k
        mode = c.mode if getattr(c, "mode", None) is not None else default_mode
        ks_used.add(k)
        hits = store.recall(c.query, k=k, mode=mode, touch=False, **recall_kwargs)
        ids = [m.id for m, *_ in hits]
        m = {
            "query": c.query,
            "k": k,
            "recall_at_k": recall_at_k(ids, c.relevant_ids, k),
            "precision_at_k": precision_at_k(ids, c.relevant_ids, k),
            "rr": reciprocal_rank(ids, c.relevant_ids),
            "ap": average_precision(ids, c.relevant_ids),
            "ndcg_at_k": ndcg_at_k(ids, c.relevant_ids, k),
            "retrieved": ids,
        }
        per.append(m)
        rs += m["recall_at_k"]; ps += m["precision_at_k"]
        rr += m["rr"]; ap += m["ap"]; ng += m["ndcg_at_k"]
    n = len(cases)
    denom = n or 1
    # Label the result with the k actually used when uniform, else -1 ("mixed").
    result_k = next(iter(ks_used)) if len(ks_used) == 1 else -1
    return EvalResult(
        n=n, k=result_k,
        recall_at_k=rs / denom, precision_at_k=ps / denom,
        mrr=rr / denom, map=ap / denom, ndcg_at_k=ng / denom,
        per_case=per,
    )

File: go/internal/eval/eval.go
// Package eval scores retrieval quality against a gold set.
//
// A port of Python's ai_houkai/eval.py, metric-for-metric, so the two ports
// report the same numbers for the same store and gold set. Dependency-free,
// matching the project's local-first philosophy.
//
// The metric functions also work standalone on any ranked list of ids:
//
//	RecallAtK([]string{"a", "b", "c"}, []string{"b"}, 2)  // -> 1.0
package eval

import (
	"context"
	"math"
	"sort"
)

// Recaller is the subset of MemoryStore that Evaluate needs. Keeping it an
// interface lets tests score a canned ranking without a live store.
type Recaller interface {
	Recall(ctx context.Context, query string, k int, opts RecallOpts) ([]Hit, error)
}

// RecallOpts and Hit mirror the store's shapes without importing it, so the
// eval package stays a leaf (memory imports nothing from here).
type RecallOpts struct {
	Mode           string
	Fusion         string
	Graph          *float32
	Diversity      *float32
	DedupThreshold *float32
	MinCosine      *float32
	ExpandRerank   bool
}

// Hit is one ranked result; only the id matters for ranking metrics.
type Hit struct {
	ID string
}

// RecallAtK is the fraction of relevant ids appearing in the top k.
// Counts distinct matches so a duplicated retrieved id cannot exceed 1.0.
func RecallAtK(retrieved []string, relevant []string, k int) float64 {
	rel := toSet(relevant)
	if len(rel) == 0 {
		return 0
	}
	top := toSet(head(retrieved, k))
	hits := 0
	for id := range top {
		if rel[id] {
			hits++
		}
	}
	return float64(hits) / float64(len(rel))
}

// PrecisionAtK is the fraction of the top k that is relevant.
func PrecisionAtK(retrieved []string, relevant []string, k int) float64 {
	rel := toSet(relevant)
	top := head(retrieved, k)
	if len(top) == 0 {
		return 0
	}
	hits := 0
	for _, id := range top {
		if rel[id] {
			hits++
		}
	}
	return float64(hits) / float64(len(top))
}

// ReciprocalRank is 1/rank of the first relevant id (0 if none retrieved).
func ReciprocalRank(retrieved []string, relevant []string) float64 {
	rel := toSet(relevant)
	for i, id := range retrieved {
		if rel[id] {
			return 1.0 / float64(i+1)
		}
	}
	return 0
}

// AveragePrecision credits each relevant id once, at its first occurrence.
func AveragePrecision(retrieved []string, relevant []string) float64 {
	rel := toSet(relevant)
	if len(rel) == 0 {
		return 0
	}
	seen := map[string]bool{}
	hits := 0
	total := 0.0
	for i, id := range retrieved {
		if rel[id] && !seen[id] {
			seen[id] = true
			hits++
			total += float64(hits) / float64(i+1)
		}
	}
	return total / float64(len(rel))
}

// DCGAtK is discounted cumulative gain (binary relevance) over the top k.
// A relevant id is credited once so duplicates cannot inflate the gain.
func DCGAtK(retrieved []string, relevant []string, k int) float64 {
	rel := toSet(relevant)
	seen := map[string]bool{}
	total := 0.0
	for i, id := range head(retrieved, k) {
		if rel[id] && !seen[id] {
			seen[id] = true
			total += 1.0 / math.Log2(float64(i+2))
		}
	}
	return total
}

// NDCGAtK is normalised DCG@k (binary relevance) in [0, 1].
func NDCGAtK(retrieved []string, relevant []string, k int) float64 {
	rel := toSet(relevant)
	if len(rel) == 0 {
		return 0
	}
	ideal := len(rel)
	if k < ideal {
		ideal = k
	}
	idcg := 0.0
	for i := 1; i <= ideal; i++ {
		idcg += 1.0 / math.Log2(float64(i+1))
	}
	if idcg == 0 {
		return 0
	}
	return DCGAtK(retrieved, relevant, k) / idcg
}

// Case is one query with its known-relevant memory ids. K and Mode are zero /
// empty by default so they fall back to Evaluate's defaults; set them per case
// only to override.
type Case struct {
	Query       string   `json:"query"`
	RelevantIDs []string `json:"relevant_ids"`
	K           int      `json:"k,omitempty"`
	Mode        string   `json:"mode,omitempty"`
}

// CaseResult is the per-case breakdown.
type CaseResult struct {
	Query        string   `json:"query"`
	K            int      `json:"k"`
	RecallAtK    float64  `json:"recall_at_k"`
	PrecisionAtK float64  `json:"precision_at_k"`
	RR           float64  `json:"rr"`
	AP           float64  `json:"ap"`
	NDCGAtK      float64  `json:"ndcg_at_k"`
	Retrieved    []string `json:"retrieved"`
}

// Result is the aggregate over all cases. K is -1 when the cases used mixed k.
type Result struct {
	N            int          `json:"n"`
	K            int          `json:"k"`
	RecallAtK    float64      `json:"recall_at_k"`
	PrecisionAtK float64      `json:"precision_at_k"`
	MRR          float64      `json:"mrr"`
	MAP          float64      `json:"map"`
	NDCGAtK      float64      `json:"ndcg_at_k"`
	PerCase      []CaseResult `json:"per_case"`
}

// Options configures a run.
type Options struct {
	DefaultK    int    // default 5
	DefaultMode string // default "hybrid"
	Recall      RecallOpts
}

// Evaluate runs each case through the store and aggregates ranking metrics.
//
// Recall is invoked read-only (the caller's RecallOpts are forwarded with
// touch disabled by the adapter) so evaluating does not perturb access-count
// or recency.
func Evaluate(ctx context.Context, store Recaller, cases []Case, opts Options) (Result, error) {
	if opts.DefaultK == 0 {
		opts.DefaultK = 5
	}
	if opts.DefaultMode == "" {
		opts.DefaultMode = "hybrid"
	}

	var sumRecall, sumPrecision, sumRR, sumAP, sumNDCG float64
	perCase := make([]CaseResult, 0, len(cases))
	ksUsed := map[int]bool{}

	for _, c := range cases {
		k := c.K
		if k == 0 {
			k = opts.DefaultK
		}
		mode := c.Mode
		if mode == "" {
			mode = opts.DefaultMode
		}
		ksUsed[k] = true

		ro := opts.Recall
		ro.Mode = mode
		hits, err := store.Recall(ctx, c.Query, k, ro)
		if err != nil {
			return Result{}, err
		}
		ids := make([]string, len(hits))
		for i, h := range hits {
			ids[i] = h.ID
		}

		cr := CaseResult{
			Query:        c.Query,
			K:            k,
			RecallAtK:    RecallAtK(ids, c.RelevantIDs, k),
			PrecisionAtK: PrecisionAtK(ids, c.RelevantIDs, k),
			RR:           ReciprocalRank(ids, c.RelevantIDs),
			AP:           AveragePrecision(ids, c.RelevantIDs),
			NDCGAtK:      NDCGAtK(ids, c.RelevantIDs, k),
			Retrieved:    ids,
		}
		perCase = append(perCase, cr)
		sumRecall += cr.RecallAtK
		sumPrecision += cr.PrecisionAtK
		sumRR += cr.RR
		sumAP += cr.AP
		sumNDCG += cr.NDCGAtK
	}

	n := len(cases)
	denom := float64(n)
	if n == 0 {
		denom = 1
	}
	// Label the result with the k actually used when uniform, else -1 ("mixed").
	resultK := -1
	if len(ksUsed) == 1 {
		keys := make([]int, 0, 1)
		for k := range ksUsed {
			keys = append(keys, k)
		}
		sort.Ints(keys)
		resultK = keys[0]
	}

	return Result{
		N: n, K: resultK,
		RecallAtK:    sumRecall / denom,
		PrecisionAtK: sumPrecision / denom,
		MRR:          sumRR / denom,
		MAP:          sumAP / denom,
		NDCGAtK:      sumNDCG / denom,
		PerCase:      perCase,
	}, nil
}

func toSet(items []string) map[string]bool {
	out := make(map[string]bool, len(items))
	for _, s := range items {
		out[s] = true
	}
	return out
}

func head(items []string, k int) []string {
	if k < 0 {
		k = 0
	}
	if k > len(items) {
		k = len(items)
	}
	return items[:k]
}
